File: 0907-koko-eating-bananas/0907-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        // for clarity refer this - https://leetcode.com/problems/koko-eating-bananas/solutions/1703687/java-c-a-very-very-well-detailed-explanation
        int left = 1;
        int right = 1000000000;
        while(left<=right){
            int mid = right+(left-right)/2;
            if(helper(piles,mid,h)){
                right = mid-1;
            }
            else{
                left = mid+1;
            }
        }
        return left;
    }
    private boolean helper(int[] piles,int k,int h){
        int hours = 0;
        for(int pile : piles){
            int div = pile/k;
            hours +=div;
            if(pile%k != 0){
                hours++;
            }
        }
        return hours<=h;
    }
}

File: 0907-koko-eating-bananas/0907-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        #for clarity refer this - https://leetcode.com/problems/koko-eating-bananas/solutions/1703687/java-c-a-very-very-well-detailed-explanation

        def helper(piles:List[int],k:int,h:int) ->bool:
            hours = 0
            for pile in piles:
                div = pile//k
                hours +=div
                if pile%k != 0:
                    hours+=1
            return hours<=h

        left = 1
        right = 1000000000
        while left<=right:
            mid = right +(left-right)//2
            if helper(piles,mid,h):
                right = mid-1
            else:
                left = mid+1
        return left
        